### PR TITLE
rdkafkacpp: Eliminate "return by const value" pessimizations.

### PR DESCRIPTION
**Pull request by Quuxplusone**
Returning a `const string` means "My caller can have a copy of
this string, but they aren't allowed to modify it" -- for example,
they aren't allowed to use move semantics to move-out-of the
returned string. Thus `x = y.name();` will use copy assignment
rather than move assignment. This is a needless pessimization
in C++11 and later.

See also the errata marked "nxd" in
https://www.aristeia.com/BookErrata/ec++3e-errata.html

### CODE REVIEW

**Review by edenhill — COMMENTED** [review]
Is this an API and ABI safe change to make?  Probably doesn't affect the API, unless people downgrade to the const-version. ABI wise?

**Comment by Quuxplusone** [comment]
> Is this an [...] ABI safe change to make?

Yes, as far as I know (and I'm fairly confident). Adding or removing `const` from the return type doesn't change the mangling on Itanium ABI (because the return type is not mangled unless it's dependent), and also doesn't change the mangling on MSVC as far as I can tell. https://godbolt.org/z/4c1ear6M9

Obviously this will change observable behavior for someone who's doing something like `decltype(Handle().name())`, but that's true of pretty much any change, because C++. :)

It certainly won't change the vtable layout or anything like that.

**Comment by Quuxplusone** [comment]
Gentle ping!

**Comment by edenhill** [comment]
Can you add a line to the 1.9.3 release's Enhancements list in CHANGELOG.md and mention this allows move optimization?

**Comment by edenhill** [comment]
I believe this will break the ABI if const is part of the symbol name mangling, so what about we just add non-const alternatives?

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,7 @@ librdkafka v1.9.3 is a maintenance release:
 ## Enhancements
 
  * Bundled zlib upgraded to version 1.2.13.
+ * The C++ API no longer returns strings "by const value," which enables better move optimization in callers.
```
Comment (edenhill) [inline CHANGELOG.md:18]:
```suggestion
 * The C++ API no longer returns strings "by const value" which enables better move optimization in callers.
```

**Review by edenhill — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (Quuxplusone) [reply CHANGELOG.md:18]: Well, we want to make it (grammatically) clear that it's the _removal_ of "return by const value" that enables the better optimization. We're not removing something _which enables optimization._ We're removing something, _which enables optimization._

Could change to `* To enable better move optimization in callers, the C++ API no longer returns strings "by const value."` if you like that better.

Comment (edenhill) [reply CHANGELOG.md:18]: Or simply remove the quotes, I think that's probably the best option.

**Comment by edenhill** [comment]
Thank you!

**Pull request merged**